### ISSUE TITLE
Reimplement, About "Algorithm 1" in your paper

### ISSUE THREAD

**Issue by guvcolie** [JiahuiYu/generative_inpainting#118]
Hi, thank you for your code!
I haved reimplement your network which separates core-code and neuralgym lib, the training code is based on Algorithm 1 in your excellent paper.
![2018-08-23 19 27 07](https://user-images.githubusercontent.com/16778790/44523316-03ccce00-a70c-11e8-802c-da0cb936d98d.png)
In my understanding， it means that firstly the discriminator is trained for 5 iters, then the generator is trained once, loop in this way. Is that right? 

The training code is below:
`           # batch_queue is sample batch fifo
            images = batch_queue.dequeue()
            _ = Loss.build_graph_with_losses(images, FLAGS)
            g_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, 'inpaint_net')
            d_vars = tf.get_collection(tf.GraphKeys.TRAINABLE_VARIABLES, 'discriminator')

            losses = Loss.build_graph_with_losses(images, FLAGS, reuse=True)
            g_loss = losses['g_loss']
            d_loss = losses['d_loss']

            learning_rate = _configure_learning_rate(FLAGS.num_samples, global_step)
            g_optimizer = _configure_optimizer(learning_rate)
            d_optimizer = g_optimizer

            g_target = g_optimizer.minimize(g_loss, global_step=global_step, var_list=g_vars)
            d_target = d_optimizer.minimize(d_loss, global_step=global_step, var_list=d_vars)

            g_update_op_list = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
            g_update_op_list.append(g_target)
            if g_update_op_list:
                updates = tf.group(*g_update_op_list)
                with tf.control_dependencies([updates]):
                    g_loss = tf.identity(g_loss, name='g_loss')
            d_update_op_list = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
            d_update_op_list.append(d_target)
            if d_update_op_list:
                updates = tf.group(*d_update_op_list)
                with tf.control_dependencies([updates]):
                    d_loss = tf.identity(d_loss, name='d_loss')

        .....

                d_iter = 0
                for batch_index in xrange(FLAGS.max_number_of_steps):
                    start_time = time()
                    if d_iter < FLAGS.D_ITER:
                        # train descriminator
                        d_iter += 1
                        discriminator_loss, count, current_lr = sess.run([d_loss, global_step, learning_rate])
                        end_time = time()
                        print('step:%05d D_loss:%.5f time_cost:%.4f' % (count, discriminator_loss, end_time - start_time))
                    else:
                        # train the generator
                        d_iter = 0
                        generative_loss, count, current_lr = sess.run([g_loss, global_step, learning_rate])
                        end_time = time()
                        print('==> step:%05d G_loss:%.5f time_cost:%.4f' % (count, generative_loss, end_time - start_time))
`
And the output is like this, do you think it's correct?  thanks!

step:00001 D_loss:18.17612 time_cost:14.1205
step:00002 D_loss:17.93037 time_cost:12.5227
step:00003 D_loss:17.72539 time_cost:12.6372
step:00004 D_loss:16.56054 time_cost:12.7271
step:00005 D_loss:15.68778 time_cost:12.7708
==> step:00006 G_loss:1.60540 time_cost:56.7784
step:00007 D_loss:15.38677 time_cost:13.1365
step:00008 D_loss:15.38883 time_cost:12.4622
step:00009 D_loss:12.38719 time_cost:13.7129
step:00010 D_loss:10.36194 time_cost:12.3818
step:00011 D_loss:-2.97061 time_cost:13.4681
==> step:00012 G_loss:1.03093 time_cost:27.2360
step:00013 D_loss:10.91956 time_cost:13.7071
step:00014 D_loss:10.10634 time_cost:13.7555
step:00015 D_loss:6.00499 time_cost:13.6784
step:00016 D_loss:-8.38905 time_cost:13.6509
step:00017 D_loss:-22.11833 time_cost:13.9234
==> step:00018 G_loss:1.02747 time_cost:28.3925
step:00019 D_loss:78.22955 time_cost:13.6580
step:00020 D_loss:51.37152 time_cost:13.8747
step:00021 D_loss:33.22484 time_cost:13.7412
step:00022 D_loss:21.71333 time_cost:13.7939
step:00023 D_loss:31.06818 time_cost:13.6712
==> step:00024 G_loss:0.96862 time_cost:28.2964
step:00025 D_loss:28.41310 time_cost:13.9727
step:00026 D_loss:26.28073 time_cost:13.7324
step:00027 D_loss:21.02436 time_cost:14.3333
step:00028 D_loss:17.93679 time_cost:14.3002
step:00029 D_loss:15.28124 time_cost:13.7549
==> step:00030 G_loss:0.51390 time_cost:28.5788
step:00031 D_loss:5.08658 time_cost:13.8227
step:00032 D_loss:8.47297 time_cost:14.4835
step:00033 D_loss:2.40379 time_cost:13.9880
step:00034 D_loss:-5.77701 time_cost:13.7595
step:00035 D_loss:-23.29973 time_cost:13.6249
==> step:00036 G_loss:0.76133 time_cost:28.4427
step:00037 D_loss:1.38272 time_cost:13.7218
step:00038 D_loss:-7.99489 time_cost:13.6285
step:00039 D_loss:-2.82512 time_cost:13.9723
step:00040 D_loss:-1.76494 time_cost:13.7711
step:00041 D_loss:20.16182 time_cost:14.4152
==> step:00042 G_loss:0.71855 time_cost:28.4300
step:00043 D_loss:9.34317 time_cost:13.6513
step:00044 D_loss:24.80732 time_cost:13.9349
step:00045 D_loss:1.21541 time_cost:13.7895
step:00046 D_loss:10.25472 time_cost:14.4245
step:00047 D_loss:5.82635 time_cost:14.2184
==> step:00048 G_loss:0.60195 time_cost:28.1839
step:00049 D_loss:-3.02651 time_cost:13.8890
step:00050 D_loss:-11.74504 time_cost:13.8308
step:00051 D_loss:-0.14545 time_cost:14.4897
step:00052 D_loss:-8.71203 time_cost:14.2027

**Comment by JiahuiYu** [comment]
@guvcolie Thanks for your interest. To separate neuralgym lib, you can simply copy related code in neuralgym to this repo, and use your own training loop. 

" it means that firstly the discriminator is trained for 5 iters, then the generator is trained once, loop in this way. Is that right?"
Correct.

**Comment by guvcolie** [comment]
the output from discriminator is negative sometimes, is that normal?

**Comment by jhyume** [comment]
@guvcolie Yes it is normal.

**Comment by JiahuiYu** [comment]
@guvcolie Yes it is normal.
